Grbl_Esp32/src/Pins/GPIOPinDetail.cpp

```diff
@@ -0,0 +1,176 @@
+        // RAII is very important here! If we throw an exception in the constructor, the resources
+        // that were allocated by the constructor up to that point _MUST_ be freed! Otherwise, you
+        // WILL get into trouble.
+
+        Assert(_capabilities != PinCapabilities::None, "Bad GPIO number");
+
+        // User defined pin capabilities
+        for (auto opt : options) {
+            if (opt.is("pu")) {
+                _attributes = _attributes | PinAttributes::PullUp;
+            } else if (opt.is("pd")) {
+                _attributes = _attributes | PinAttributes::PullDown;
+            } else if (opt.is("low")) {
+                _attributes = _attributes | PinAttributes::ActiveLow;
+            } else if (opt.is("high")) {
+                // Default: Active HIGH.
+            } else {
+                Assert(false, "Bad GPIO option passed to pin %d: %s", int(index), opt());
+            }
+        }
+
+        // Update the R/W mask for ActiveLow setting
+        if (_attributes.has(PinAttributes::ActiveLow)) {
+            _readWriteMask = HIGH;
+        } else {
+            _readWriteMask = LOW;
+        }
+    }
+
+    PinCapabilities GPIOPinDetail::capabilities() const { return _capabilities; }
+
+    void GPIOPinDetail::write(int high) {
+        Assert(_attributes.has(PinAttributes::Output), "Pin has no output attribute defined. Cannot write to it.");
+        int value = _readWriteMask ^ high;
+        __digitalWrite(_index, value);
+    }
+    int GPIOPinDetail::read() {
+        auto raw = __digitalRead(_index);
+        return raw ^ _readWriteMask;
+    }
+
+    void GPIOPinDetail::setAttr(PinAttributes value) {
+        // These two assertions will fail if we do them for index 1/3 (Serial uart). This is because
+        // they are initialized by HardwareSerial well before we start our main operations. Best to
+        // just ignore them for now, and figure this out later. TODO FIXME!
+
+        // Check the attributes first:
+        Assert(value.validateWith(this->_capabilities) || _index == 1 || _index == 3,
+               "The requested attributes don't match the pin capabilities");
+        Assert(!_attributes.conflictsWith(value) || _index == 1 || _index == 3,
+               "Attributes on this pin have been set before, and there's a conflict.");
+
+        _attributes = value;
+
+        // Handle attributes:
+        uint8_t pinModeValue = 0;
+
+        if (value.has(PinAttributes::Input)) {
+            pinModeValue |= INPUT;
+        } else if (value.has(PinAttributes::Output)) {
+            pinModeValue |= OUTPUT;
+        }
+
+        // PU/PD should be specified by the user. Code has nothing to do with them:
+        if (_attributes.has(PinAttributes::PullUp)) {
+            pinModeValue |= PULLUP;
+        } else if (_attributes.has(PinAttributes::PullDown)) {
+            pinModeValue |= PULLDOWN;
+        }
+
+        // If the pin is ActiveLow, we should take that into account here:
```

Comment:
According to Mitch, we should not digitalWrite for() INPUT pins

Reply:
Mitch is correct. The InitialOn flag is only relevant for output pins that need to be initialized active high (or low when inverted). Some output pins should have that to ensure there is not glitching. This code takes care of that. 

Currently it's not used, because INVERT is still all over the place.